File: src/learning_tf/scripts/turtle_tf_broadcaster.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import roslib
roslib.load_manifest('learning_tf')
import tf.transformations
import rospy

import tf
import turtlesim.msg

def handle_turtle_pose(msg, turtlename):
    br = tf.TransformBroadcaster()  # 创建tf.TransformBroadcaster()实例
    # 处理函数针对turtle的位置信息广播平移和旋转，作为变换，从“world”坐标系到“turtleX”坐标系进行发布
    br.sendTransform((msg.x, msg.y, 0),
                     tf.transformations.quaternion_from_euler(0, 0, msg.theta),
                     rospy.Time.now(),
                     turtlename,
                     "world")
    
if __name__ == '__main__':
    # 初始化ROS节点
    rospy.init_node('turtle_tf_broadcaster')  # anonymous=True:表示每个节点具有唯一名称
    turtlename = rospy.get_param('~turtle')  # rospy.get_param(param_name):从参数服务器获取值
    # 此节点接受一个名为“turtle”的单个参数
    rospy.Subscriber('/%s/pose' % turtlename,
                     turtlesim.msg.Pose,
                     handle_turtle_pose,
                     turtlename)  # rospy.Subscriber("chatter", String, callback)
    rospy.spin()  # 阻塞，直到 ROS 节点关闭。将活动让给其他线程。

File: src/learning_tf/src/turtle_tf_broadcaster.cpp
#include <ros/ros.h>
#include <tf/transform_broadcaster.h>
#include <turtlesim/Pose.h>

std::string turtle_name;  // std::string: c++中字符串类

void poseCallback(const turtlesim::PoseConstPtr& msg)
{
    // 创建tf的广播器
    static tf::TransformBroadcaster br;

    // 初始化tf数据
    tf::Transform transform;
    transform.setOrigin(tf::Vector3(msg->x, msg->y, 0.0));
    tf::Quaternion q;
    q.setRPY(0, 0, msg->theta);
    transform.setRotation(q);

    // 广播world与海龟坐标系之间的tf数据
    br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), "world", turtle_name));
}

// int main(int argc, char *argv[])
int main(int argc, char** argv)
// argc:整数,用来统计你运行程序时送给main函数的命令行参数的个数
// * argv[ ]: 指针数组，用来存放指向你的字符串参数的指针，每一个元素指向一个参数
{
    // 初始化ROS节点
    ros::init(argc, argv, "my_tf_broadcaster");

    // 输入参数作为海龟的节点
    if(argc!=2)
    {
        ROS_ERROR("need turtle name as argument");
        return -1;
    }
    
    turtle_name = argv[1];

    // 订阅海龟的位姿话题
    ros::NodeHandle node;
    ros::Subscriber sub = node.subscribe(turtle_name+"/pose", 10, &poseCallback);

    // 循环等待回调函数
    ros::spin();

    return 0;
}
